experiment/adscert/signer.go

```diff
@@ -0,0 +1,129 @@
+		return signatureMessage, nil
+	}
+	return "", fmt.Errorf("error signing request: %s", resp.GetSignatureOperationStatus())
+}
+
+func newInProcessSigner(inProcessSignerConfig config.InProcess) *inProcessSigner {
+	return &inProcessSigner{
+		signatory: signatory.NewLocalAuthenticatedConnectionsSignatory(
+			inProcessSignerConfig.Origin,
+			rand.Reader,
+			clock.New(),
+			discovery.NewDefaultDnsResolver(),
+			discovery.NewDefaultDomainStore(),
+			time.Duration(inProcessSignerConfig.DNSCheckIntervalInSeconds)*time.Second,
+			time.Duration(inProcessSignerConfig.DNSRenewalIntervalInSeconds)*time.Second,
+			[]string{inProcessSignerConfig.PrivateKey}),
+	}
+}
+
+func createRequestInfo(destinationURL string, body []byte) *api.RequestInfo {
+	reqInfo := &api.RequestInfo{}
+	signatory.SetRequestInfo(reqInfo, destinationURL, body)
+	return reqInfo
+}
+
+type remoteSigner struct {
+	signatory signatory.AuthenticatedConnectionsSignatory
+}
+
+func (rs *remoteSigner) Sign(destinationURL string, body []byte) (string, error) {
+	// The RequestInfo proto contains details about the individual ad request
+	// being signed.  A SetRequestInfo helper function derives a hash of the
+	// destination URL and body, setting these value on the RequestInfo message.
+	reqInfo := &api.RequestInfo{}
+	signatory.SetRequestInfo(reqInfo, destinationURL, []byte(body))
+
+	// Request the signature.
+	signatureResponse, err := rs.signatory.SignAuthenticatedConnection(
+		&api.AuthenticatedConnectionSignatureRequest{
+			RequestInfo: reqInfo,
+		})
+	if err != nil {
+		return "", err
+	}
+	if signatureResponse != nil && signatureResponse.SignatureOperationStatus == api.SignatureOperationStatus_SIGNATURE_OPERATION_STATUS_OK {
+		signatureMessage := signatureResponse.RequestInfo.SignatureInfo[0].SignatureMessage
+		return signatureMessage, err
+	}
+	return "", fmt.Errorf("error signing request: %s", signatureResponse.GetSignatureOperationStatus())
+}
+
+func newRemoteSigner(remoteSignerConfig config.Remote) (*remoteSigner, error) {
+	// Establish the gRPC connection that the client will use to connect to the
+	// signatory server.  This basic example uses unauthenticated connections
+	// which should not be used in a production environment.
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+	conn, err := grpc.Dial(remoteSignerConfig.Url, opts...)
+	if err != nil {
+		fmt.Errorf("failed to dial remote signer: %v", err)
+	}
+	//defer conn.Close() -- where this should be?
+
+	clientOpts := &signatory.AuthenticatedConnectionsSignatoryClientOptions{
+		Timeout: time.Duration(remoteSignerConfig.SigningTimeoutMs) * time.Millisecond}
+	signatoryClient := signatory.NewAuthenticatedConnectionsSignatoryClient(conn, clientOpts)
+	return &remoteSigner{signatory: signatoryClient}, nil
+
+}
+
+type NilSigner struct {
+}
+
+func (ns *NilSigner) Sign(destinationURL string, body []byte) (string, error) {
+	return "", nil
+}
```

Comment:
Do you think it would be a good idea to separate the in process signer, remote signer, and nil signer to their own files?

Reply:
Separated In-Process and Remote signers to separate files. NilSigner is now in the same file with interface. Should I move it too?